flatland_server/include/flatland_server/message_server.h

```diff
@@ -1,217 +1,113 @@
-    return MessageView<T>(messages_, start, end);
-  }
+  Subscriber() = default;
 
-  void add_new_message(const T& msg);
-  void clean_old_messages() override;
-};
+  explicit Subscriber(MessageTopic<T>* topic,
+                      const std::function<void(const T&)>& callback_function)
+      : topic_(topic), callback_function_(std::move(callback_function)) {}
 
-template <class T>
-class Subscriber : public SubscriberBase {
+ private:
   MessageTopic<T>* topic_;
-  ros::Time last_subscription_time;
-
-  Subscriber(MessageTopic<T>* topic)
-      : topic_(topic), last_subscription_time(ros::Time::now()) {}
-
- public:
-  Subscriber() {}
-  ~Subscriber() {}
-  MessageView<T> receive();
+  std::function<void(const T&)> callback_function_;
 
-  friend class MessageServer;
+  friend class Publisher<T>;
 };
 
 template <class T>
-class Publisher : public PublisherBase {
-  MessageTopic<T>* topic_;
-
-  Publisher(MessageTopic<T>* topic) : topic_(topic) {}
-
+class Publisher {
  public:
-  Publisher() {}
-  ~Publisher() {}
+  Publisher() = default;
+  explicit Publisher(MessageTopic<T>* topic) : topic_(topic) {}
+
   void publish(const T&);
 
-  friend class MessageServer;
+ private:
+  MessageTopic<T>* topic_;
 };
 
 class MessageServer {
- public:
-  std::unordered_map<std::string, std::unique_ptr<MessageTopicBase>>
-      messageTopics_;
+ private:
+  std::unordered_map<std::string, std::unique_ptr<MessageTopicBase>> topics_;
 
   template <class T>
-  MessageTopic<T>* create_topic(
-      const std::string& name,
-      const ros::Duration& message_life = ros::Duration(1));
+  MessageTopic<T>* get_message_topic(const std::string& name);
 
+ public:
   template <class T>
   Subscriber<T> subscribe(
       const std::string& name,
-      const ros::Duration& message_lifetime = ros::Duration(1));
+      const std::function<void(const T&)>& callback_function);
 
   template <class T>
-  Publisher<T> advertise(
-      const std::string& name,
-      const ros::Duration& message_lifetime = ros::Duration(1));
-
-  void clean_old_topics();
+  Publisher<T> advertise(const std::string& name);
 };
 
 ///
 /// Start of Implementation section
 ///
 
 template <class T>
-MessageTopic<T>* MessageServer::create_topic(
-    const std::string& name, const ros::Duration& message_life) {
-  flatland_server::MessageTopic<T>* topic;
-  auto msgTopic = messageTopics_.find(name);
-  if (msgTopic == messageTopics_.end()) {
-    topic = new flatland_server::MessageTopic<T>(message_life);
-    messageTopics_.emplace(name, std::unique_ptr<MessageTopic<T>>(topic));
-    ROS_INFO_STREAM("Flatland Server topic created: " << name);
-  } else {
-    topic = static_cast<MessageTopic<T>*>((msgTopic->second).get());
+MessageTopic<T>* MessageServer::get_message_topic(const std::string& name) {
+  auto msg_topic = topics_.find(name);
+  if (msg_topic != topics_.end()) {
+    return static_cast<MessageTopic<T>*>((msg_topic->second).get());
```

Comment:
You can probably simplify here

if (msg_topic == topics_.end()) {
    msg_topic = ...emplace(...).first;
}
return ...;

Reply:
I think it should be the same logic here, I have to return the newly added one in the if function as well. I just did this so its an early return rather than putting more stuff in the if statement